### ISSUE TITLE
Reduce gradients problem of DataParallelTable:accGradParameters()

### ISSUE THREAD

**Issue by Jerrynet** [torch/cunn#209]
At [DataParallelTable:__backward()#L186](https://github.com/torch/cunn/blob/97f66e19a3b9ce30e4717164b8eaea257b569bf8/DataParallelTable.lua#L186)
```DataParallelTable:accGradParameters()``` calls ```DataParallelTable:__backward()``` and it sums gradients to the base GPU more than once if a user does not run ```nn.Module:zeroGradParameters()``` at every iteration.
In my opinion, gradients should be reduced only once after a user called ```DataParallelTable:backward()``` or ```DataParallelTable:accGradParameters()``` several times.

Any ideas are welcome.

**Comment by soumith** [comment]
@Jerrynet i think this is doing exactly what it says on the can. :backward is supposed to call updateGradInput + accGradParameters, no complex logic there.

**Comment by Jerrynet** [comment]
@soumith I think it accumulates gradients which are already accumulated if I call ```accGradParameters()``` again?
For example, two GPUs: GPU1, GPU2
first call ```accGradParameters()```:
```self.flattenedParams ``` = ```grad_gpu1_firstCall+grad_gpu2_firstCall```
second call ```accGradParameters()```:
```self.flattenedParams ``` = ```(grad_gpu1_firstCall+grad_gpu2_firstCall) + [grad_gpu1_secondCall+(grad_gpu2_firstCall+grad_gpu2_secondCall)]```

**Comment by soumith** [comment]
yes, accGradParameters computes gradParameters and accumulates them. That's the definition of the function. It does not do anything fancy like keeping track of what you want to do...

**Comment by Jerrynet** [comment]
I think it should be 
```self.flattenedParams``` = ```(grad_gpu1_firstCall+grad_gpu2_firstCall) + grad_gpu1_secondCall+grad_gpu2_secondCall``` after the second call, 
instead of 
```self.flattenedParams``` = ```(grad_gpu1_firstCall+grad_gpu2_firstCall) + [grad_gpu1_secondCall+(grad_gpu2_firstCall+grad_gpu2_secondCall)]```?

**Comment by soumith** [comment]
@Jerrynet ok i now see what you are saying, writing it as a formula made it much clearer :)
Yea looks like a bug for @colesbury to fix

**Comment by Jerrynet** [comment]
@soumith Sorry for my expressions, and thanks for your patience :)

**Comment by colesbury** [comment]
Thanks @Jerrynet for the bug report!
